engine/access/subscription/block_tracker.go

```diff
@@ -0,0 +1,128 @@
+package subscription
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/counters"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/state/protocol"
+
+	"github.com/onflow/flow-go/storage"
+)
+
+// GetHighestHeight is a function type for getting the highest height.
+// Block status could be only BlockStatusSealed or BlockStatusFinalized.
+type GetHighestHeight func(flow.BlockStatus) (uint64, error)
+
+// BlockTracker is an interface for tracking blocks and handling block-related operations.
+type BlockTracker interface {
+	BaseTracker
+	// GetHighestHeight returns the highest height based on the specified block status which could be only BlockStatusSealed
+	// or BlockStatusFinalized.
+	// No errors are expected during normal operation.
+	GetHighestHeight(flow.BlockStatus) (uint64, error)
+	// ProcessOnFinalizedBlock drives the subscription logic when a block is finalized.
+	// The input to this callback is treated as trusted. This method should be executed on
+	// `OnFinalizedBlock` notifications from the node-internal consensus instance.
+	ProcessOnFinalizedBlock() error
+}
+
+var _ BlockTracker = (*BlockTrackerImpl)(nil)
+
+// BlockTrackerImpl is an implementation of the BlockTracker interface.
+type BlockTrackerImpl struct {
+	BaseTracker
+	state       protocol.State
+	broadcaster *engine.Broadcaster
+
+	// finalizedHighestHeight contains the highest consecutive block height for which we have received a new notification.
+	finalizedHighestHeight counters.StrictMonotonousCounter
+	// sealedHighestHeight contains the highest consecutive block height for which we have received a new notification.
+	sealedHighestHeight counters.StrictMonotonousCounter
+}
+
+// NewBlockTracker creates a new BlockTrackerImpl instance.
+//
+// Parameters:
+// - state: The protocol state used for retrieving block information.
+// - rootHeight: The root block height, serving as the baseline for calculating the start height.
+// - headers: The storage headers for accessing block headers.
+// - broadcaster: The engine broadcaster for publishing notifications.
+//
+// Returns:
```

Comment:
No errors are expected during normal operations, usually we describe errors where it's part of the flow and can be represented as sentinel. This is not the case for this method, I suggest to update the godoc.